internal/command/jsonfunction/function.go

```diff
@@ -0,0 +1,95 @@
+package jsonfunction
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+// FormatVersion represents the version of the json format and will be
+// incremented for any change to this format that requires changes to a
+// consuming parser.
+const FormatVersion = "1.0"
+
+// functions is the top-level object returned when exporting function signatures
+type functions struct {
+	FormatVersion string                        `json:"format_version"`
+	Signatures    map[string]*FunctionSignature `json:"function_signatures,omitempty"`
+}
+
+// FunctionSignature represents a function signature.
+type FunctionSignature struct {
+	// Description is an optional human-readable description
+	// of the function
+	Description string `json:"description,omitempty"`
+
+	// ReturnTypes is the ctyjson representation of the function's
+	// return types based on supplying all parameters using
+	// dynamic types. Functions can have dynamic return types.
+	// TODO? could we use cty.Type here instead of calling ctyjson.MarshalType manually?
+	// TODO? see: https://github.com/zclconf/go-cty/blob/main/cty/json/type.go
+	ReturnType json.RawMessage `json:"return_type"`
+
+	// Parameters describes the function's fixed positional parameters.
+	Parameters []*parameter `json:"parameters,omitempty"`
+
+	// VariadicParameter describes the function's variadic
+	// parameters, if any are supported.
+	VariadicParameter *parameter `json:"variadic_parameter,omitempty"`
+}
+
+func newFunctions() *functions {
+	signatures := make(map[string]*FunctionSignature)
+	return &functions{
+		FormatVersion: FormatVersion,
+		Signatures:    signatures,
+	}
+}
+
+func Marshal(f map[string]function.Function) ([]byte, error) {
+	signatures := newFunctions()
+
+	for k, v := range f {
+		signature, err := marshalFunction(v)
+		if err != nil {
+			fmt.Printf("Failed to serialize function %q: %s\n", k, err) // TODO! handle error
+			continue
```

Comment:
This should `return nil, err` so the error can be correctly handled in `command/metadata_functions.go`.

Reply:
Actually, it may be better to collect up all the functions that have failed to marshal and have `Marshal()` return them in one well-formatted error.

Reply:
I collect all the errors now and output them using `showDiagnostics`

<img width="539" alt="CleanShot 2023-01-17 at 12 06 48@2x" src="https://user-images.githubusercontent.com/45985/212883544-cd4b5714-f2dd-469d-9488-f289d1a9c1e6.png">